File: cmd/graph/1311/1311.py
# 1311. Get Watched Videos by Your Friends

from queue import Queue

class Solution:
    def watchedVideosByFriends(self, watchedVideos: List[List[str]], friends: List[List[int]], id: int, level: int) -> List[str]:
        dict_freq = {}
        lev = 0
        fr_set = set()
        visited = [False]*len(friends)

        q = Queue()
        q.put(id)

        # find level friends
        while True:
            size = q.qsize()
            for i in range(size):
                fr_id = q.get()
                for fr in friends[fr_id]:
                    if not visited[fr]:
                        q.put(fr)
                        visited[fr] = True
            lev = lev + 1
            if level == lev:
                break

        # get unique friends id
        while not q.empty():
            fr = q.get()
            if fr != id:
                fr_set.add(fr)

        # count watched video frequency
        for fr in fr_set:
            for mov in watchedVideos[fr]:
                dict_freq[mov] = dict_freq.get(mov, 0) + 1

        # sort result
        sorts = sorted(dict_freq.items(), key=lambda item: (item[1], item[0]))

        # fill result
        return [k for k, v in sorts]

File: cmd/graph/1311/1311.go
package _1311

import "sort"

type movieStatistic struct {
	key   string
	count int
}

// 1311. Get Watched Videos by Your Friends

func watchedVideosByFriends(watchedVideos [][]string, friends [][]int, id int, level int) []string {
	queue := traverseToTargetLevel(friends, id, level)
	friendSet := getUniqueFriendsId(queue, id)
	frequency := calculateMovieFrequency(watchedVideos, friendSet)
	movieStatistics := buildMovieStatistic(frequency)
	return fillResult(movieStatistics)
}

func traverseToTargetLevel(friends [][]int, id int, targetLevel int) []int {
	currentLevel := 0
	visited := make([]bool, len(friends))

	q := make([]int, 0, 100)
	q = append(q, id)

	for true {
		size := len(q)
		for i := 0; i < size; i++ {
			friendId := q[0]
			q = q[1:]
			for _, frId := range friends[friendId] {
				if !visited[frId] {
					q = append(q, frId)
					visited[frId] = true
				}
			}
		}
		currentLevel++
		if targetLevel == currentLevel {
			break
		}
	}
	return q
}

func getUniqueFriendsId(q []int, id int) map[int]bool {
	friendSet := make(map[int]bool)

	for len(q) > 0 {
		friendId := q[0]
		q = q[1:]
		if friendId != id {
			if _, ok := friendSet[friendId]; !ok {
				friendSet[friendId] = true
			}
		}
	}
	return friendSet
}

func calculateMovieFrequency(watchedVideos [][]string, friendSet map[int]bool) map[string]int {
	frequency := make(map[string]int, len(watchedVideos))
	for friendId := range friendSet {
		for _, movie := range watchedVideos[friendId] {
			frequency[movie]++
		}
	}
	return frequency
}

func buildMovieStatistic(frequency map[string]int) []*movieStatistic {
	movieStatistics := make([]*movieStatistic, 0, len(frequency))
	for movie, freq := range frequency {
		movieStatistics = append(movieStatistics, &movieStatistic{key: movie, count: freq})
	}

	sort.Slice(movieStatistics, func(i, j int) bool {
		left := movieStatistics[i]
		right := movieStatistics[j]
		if left.count != right.count {
			return left.count < right.count
		}
		return left.key < right.key
	})

	return movieStatistics
}

func fillResult(tmp []*movieStatistic) []string {
	var result []string
	for _, movieStat := range tmp {
		result = append(result, movieStat.key)
	}
	return result
}
